File: cpp/GNEO/test/test_gptneo.cpp
#include "test_common.hpp"
#include "GPTNeo.hpp"
#include "tokenizer.hpp"

using namespace mnn;
using namespace mtb;

#define FP_T float


void test_GPTNeo(){
    START_TEST();
    GPTNeoModel<FP_T> model(4, 768);
    // Load the model state dict
    model.load_model("../model_state_dict");  
    
    Tensor<int> input_ids({1, 5}, {7454, 2402,257,640,11});
    Tensor<int> position_ids({1, 5}, {0, 1, 2, 3, 4});
    auto output = load_data<FP_T>(
        "build/test/temp/output.bin", 1* 5* 50257);

    Tensor<FP_T> output_t = model.forward(input_ids, position_ids);

    // Check the 
    compare_data<FP_T>(
        output_t.data().get(), output.data(), 
        output.size(), 1e-5);

    auto next_token = output_t.slice(
        {{0,1}, {4,5}, {0,50257}});
    
    std::cout << "shape of next token: "
              << next_token.shape() << 
              " size: " << next_token.size() 
              << " stride: " << next_token.strides()
              << std::endl;

    auto next_token_index = mtb::argmax(next_token, -1);
    std::cout << "next token index shape: "
              << next_token_index.shape() << std::endl;
    
    assert(next_token_index.size() == 1);
    assert(next_token_index.data()[0] == 612);

    Tokenizer tokenizer("../python/tokenizer/vocab.json");
    tokenizer.parse_vocab();

    auto next_token_str = tokenizer.get_token(
        next_token_index.data()[0]
    );
    std::cout << "next token: " << next_token_str << std::endl;

    PASSLOG();
}

int main(int argc, char** argv) {
    test_GPTNeo();
    return 0;
}

File: cpp/GNEO/test/test_gptneo.py
import sys
# pwd 
import os
current_working_directory = os.getcwd()
print("Current working directory:", current_working_directory)

sys.path.append("../python/")  # Adjust the path as necessary
from GPTNeo import GPTNeoModel
import numpy as np


model = GPTNeoModel(4, 768)
model.load("../model_state_dict")

input_ids = np.array([[ 7454, 2402,257,640,11]]).astype(np.int32)  # Example input
position_ids = np.arange(input_ids.shape[1]).reshape(1, -1).astype(np.int32)  # Position IDs

print("Input shape:", input_ids.shape)
print("Position IDs shape:", position_ids.shape)
output = model(input_ids, position_ids=position_ids)
print("Output shape:", output.shape)
print("Output:", output)
next_token_logits = output[:, -1, :]
next_token = np.argmax(next_token_logits, axis=-1).reshape(-1, 1)
print("Next token:", next_token)


# save the output to a file
save_path = "build/test/temp"
output.tofile(os.path.join(save_path, "output.bin"))